api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -427,13 +429,21 @@ private void checkForRedundantPartitions(PartitionField field) {
       dedupFields.put(dedupKey, field);
     }
 
+    public Builder caseSensitive(boolean sensitive) {
+      this.caseSensitive = sensitive;
+      return this;
+    }
+
     public Builder withSpecId(int newSpecId) {
       this.specId = newSpecId;
       return this;
     }
 
     private Types.NestedField findSourceColumn(String sourceName) {
-      Types.NestedField sourceColumn = schema.findField(sourceName);
+      Types.NestedField sourceColumn =
+          this.caseSensitive
+              ? schema.findField(sourceName)
+              : schema.caseInsensitiveFindField(sourceName);
```

Comment:
`caseInsensitiveFindField` uses normalized lower-case string for name -> id indexing.  who can ensure the schema don't have two fields with names like `data` and `DATA`? Otherwise, `caseInsensitiveFindField` search is ambiguous. I am wondering if the `caseSensitive` config need to be pushed into the `Schema` and spec?

Reply:
Now you're pushing me into the deep end of the pool. 😉 I can't answer your question. Your question would have to be directed to someone with greater knowledge of Iceberg's case-insensitive plan.

Reply:
the question is really meant for @rdblue and @RussellSpitzer :) There is also the question of engine interoperability on case sensitivity. If the table is defined as case insensitive, ideally validation should reject the schema that @dramaticlly pointed out.

```
table {
  1: id: required int
  2: data: required string
  3: DATA: required string
}
```

Reply:
@stevenzwu @rdblue @RussellSpitzer I suppose the answer to this question is to NOT use a case-sensitive schema in a case-insensitive way. In other words, if the schema is case-sensitive and contains columns named Make and MAKE then the result of calling `caseInsensitiveFindField("Make")` is undefined.

Reply:
The current behavior is to fail when a schema that requires case sensitivity is used in a case insensitive way. Right now that would be by throwing an exception when [creating the case insensitive index](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/types/TypeUtil.java#L184-L190).

Because Iceberg needs to work in contexts that are both case sensitive and case insensitive, I think it is the right strategy to fail at runtime. Iceberg can't impose case insensitivity on engines that are case sensitive or vice versa.

In nearly all situations, this works fine and is not noticed because it is very uncommon to have both `data` and `DATA` in a schema.

Reply:
@rdblue The timing of your answer is impeccable because at the very moment I received your response I was debugging that very method, `TypeUtil.indexByLowerCaseName()`, to see how it handles `data` and `DATA`. The answer is that the method does not throw an exception at all.

If you start with this schema:
```
Schema schema = new Schema(
        required(1, "id", Types.IntegerType.get()),
        required(2, "data", Types.StringType.get()),
        required(3, "DATA", Types.StringType.get()));
```
then calling `TypeUtil.indexByLowerCaseName()` will return a RegularImmutableMap of
```
{
    "data": 3,
    "id": 1
}
```
So either the result is undefined or (by implementation detail) is the ID of the last column with case-insensitive matching name.

Reply:
I thought that `ImmutableMap.Builder` would reject duplicate keys. In any case, we should probably update the code I linked to so that it catches this and throws a nice error message.

Reply:
Yes, I am sure I'm getting `RegularImmutableMap`

<img width="848" alt="image" src="https://github.com/user-attachments/assets/38841220-bcb2-4eab-880f-4994038dc7ab">

Reply:
I updated the code to use `ImmutableMap.Builder`. It will now throw on duplicate keys. It still returns a RegularImmutableMap instance